src/hb-ot-cmap-table.hh

```diff
@@ -953,16 +947,73 @@ struct CmapSubtableFormat14
     this->format = 14;
 
     auto src_tbl = reinterpret_cast<const CmapSubtableFormat14*> (src_base);
-    c->copy_all (hb_iter (src_tbl->record),
-		 unicodes, glyphs, glyph_map, src_base, this);
+
+    // Some versions of OTS require that offsets are in order. Due to the use
+    // of push()/pop_pack() serializing the variation records in order results
+    // in the offsets being in reverse order (first record has the largest
+    // offset). While this is perfectly valid, it will cause some versions of
+    // OTS to consider this table bad.
+    //
+    // So to prevent this issue we serialize the variation records in reverse
+    // order, so that the offsets are ordered from small to large. Since
+    // variation records are supposed to be in increasing order of varSelector
+    // we then have to reverse the order of the written variation selector
+    // records after everything is finalized.
+    hb_vector_t<hb_pair_t<unsigned, unsigned>> obj_indices;
+    for (int i = src_tbl->record.len - 1; i >= 0; i--)
+    {
+      hb_pair_t<unsigned, unsigned> result =
+          src_tbl->record[i].copy (c, unicodes, glyphs, glyph_map, src_base, this);
+      if (result.first || result.second)
+        obj_indices.push (result);
+    }
 
     if (c->length () - table_initpos == CmapSubtableFormat14::min_size)
+    {
       c->revert (snap);
-    else
+      return;
+    }
+
+    int tail_len = init_tail - c->tail;
+    c->check_assign (this->length, c->length () - table_initpos + tail_len);
+    c->check_assign (this->record.len,
+                     (c->length () - table_initpos - CmapSubtableFormat14::min_size) /
+                     VariationSelectorRecord::static_size);
+
+    // Correct the incorrect write order by reversing the order of the variation
+    // records array.
+    _reverse_variation_records ();
+
+    // Now that records are in the right order, we can set up the offsets.
+    _add_links_to_variation_records (c, obj_indices);
+  }
+
+  void _reverse_variation_records ()
```

Comment:
This sounds like should be a `reverse` operation on `hb_array_t` or something similar.

Reply:
A general purpose reverse on array_t wouldn't quite work here, since we're only reversing a single field in VariationSelectorRecord and not the whole record.

Reply:
On second thought, we end up overwriting the other two fields of VariationSelector in the next method, so a general purpose reverse would work.